File: 2023/day11/main.go
package main

import (
	"bufio"
	"bytes"
	_ "embed"
	"fmt"
	"slices"
	"strconv"
	"strings"

	"github.com/samber/lo"
)

//go:embed input.txt
var input []byte

type Galaxy struct {
	Row, Col int
}

func (p *Galaxy) Dist(o *Galaxy) int {
	x := p.Col - o.Col
	if x <= 0 {
		x *= -1
	}
	y := p.Row - o.Row
	if y <= 0 {
		y *= -1
	}
	return x + y
}

/*
* part 1: 9724940
* part 2: 569052586852
 */
func main() {
	fmt.Printf("part1: %s\n", solve(input, 2))
	fmt.Printf("part2: %s\n", solve(input, 1_000_000))
}

func expand(universe [][]string, galaxies []*Galaxy, factor int) {
	factor = factor - 1 //one row already in the place
	emptyRows := make([]int, 0)
	emptyCols := make([]int, 0)

	for r, row := range universe {
		if !slices.Contains(row, "#") {
			emptyRows = append(emptyRows, r)
		}
	}

	for c := 0; c < len(universe[0]); c++ {
		empty := true
		for r := 0; r < len(universe); r++ {
			if universe[r][c] == "#" {
				empty = false
				break
			}
		}
		if empty {
			emptyCols = append(emptyCols, c)
		}
	}

	for _, galaxy := range galaxies {
		deltaRow := 0
		for _, row := range emptyRows {
			if galaxy.Row > row {
				deltaRow += factor
			}
		}
		deltaCol := 0
		for _, Col := range emptyCols {
			if galaxy.Col > Col {
				deltaCol += factor
			}
		}
		galaxy.Row += deltaRow
		galaxy.Col += deltaCol
	}
}

// solve
func solve(input []byte, factor int) string {
	scanner := bufio.NewScanner(bytes.NewBuffer(input))
	universe := make([][]string, 0)
	for scanner.Scan() {
		line := scanner.Text()
		universe = append(universe, strings.Split(line, ""))
	}

	galaxies := make([]*Galaxy, 0)
	for x, row := range universe {
		for y, char := range row {
			if char == "#" {
				galaxies = append(galaxies, &Galaxy{x, y})
			}
		}
	}

	expand(universe, galaxies, factor)

	distances := make([]int, 0)
	for i, galaxy := range galaxies {
		for j := i + 1; j < len(galaxies); j++ {
			distances = append(distances, galaxy.Dist(galaxies[j]))
		}
	}
	if err := scanner.Err(); err != nil {
		return err.Error()
	}
	return strconv.Itoa(lo.Sum(distances))
}

File: 2023/day11/main.rs
#[derive(Debug, PartialEq)]
struct Galaxy {
    row: u64,
    col: u64,
}

impl Galaxy {
    fn distances(&self, o: &Galaxy) -> u64 {
        ((self.col as i64 - o.col as i64).abs() + (self.row as i64 - o.row as i64).abs()) as u64
    }
}

fn expand(
    galaxies: Vec<Galaxy>,
    empty_rows: Vec<u64>,
    empty_cols: Vec<u64>,
    factor: u64,
) -> Vec<Galaxy> {
    let factor = factor - 1;
    galaxies
        .iter()
        .map(|galaxy| {
            let dr = empty_rows.iter().filter(|&&r| r < galaxy.row).count() as u64 * factor;
            let dc = empty_cols.iter().filter(|&&r| r < galaxy.col).count() as u64 * factor;
            Galaxy {
                row: galaxy.row + dr,
                col: galaxy.col + dc,
            }
        })
        .collect::<Vec<Galaxy>>()
}

fn distances(input: &str, factor: u64) -> u64 {
    let universe = input
        .lines()
        .map(|line| line.chars().collect::<Vec<_>>())
        .collect::<Vec<_>>();
    let galaxies: Vec<_> = universe
        .iter()
        .enumerate()
        .map(|(x, values)| {
            values
                .iter()
                .enumerate()
                .flat_map(move |(y, char)| {
                    if *char == '#' {
                        Some(Galaxy {
                            row: x as u64,
                            col: y as u64,
                        })
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>()
        })
        .flatten()
        .collect();

    let empty_rows: Vec<u64> = universe
        .iter()
        .enumerate()
        .filter(|(_x, line)| !(*line).contains(&'#'))
        .map(|(x, _line)| x as u64)
        .collect();

    let empty_cols = universe[0]
        .iter()
        .enumerate()
        .filter(|(c, _char)| {
            for r in 0..universe.len() {
                if universe[r][*c] == '#' {
                    return false;
                }
            }
            return true;
        })
        .map(|(c, _line)| c as u64)
        .collect::<Vec<u64>>();

    let galaxies = expand(galaxies, empty_rows, empty_cols, factor);

    let mut distance = 0;
    for i in 0..galaxies.len() {
        for j in i + 1..galaxies.len() {
            distance += galaxies[i].distances(&galaxies[j])
        }
    }
    distance
}

fn process_part1(input: &str) -> String {
    let distances = distances(input, 2);
    distances.to_string()
}

fn process_part2(input: &str) -> String {
    let distances = distances(input, 1_000_000);
    distances.to_string()
}

fn main() {
    let input = include_str!("./input.txt");
    println!(
        "part1: {}\npart2: {}",
        process_part1(input),
        process_part2(input)
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() -> miette::Result<()> {
        let input: &str = include_str!("./example.txt");
        assert_eq!(374, distances(input, 2));
        Ok(())
    }
    #[test]
    fn test_part2() -> miette::Result<()> {
        let input = include_str!("./example.txt");
        assert_eq!(8410, distances(input, 100));
        Ok(())
    }
}
